### PR TITLE
Fix issue 20705 - `-preview=rvaluerefparam` does not work with templa…

### PR DESCRIPTION
**Pull request by thewilsonator**
…te deduction

### CODE REVIEW

File: src/dmd/dtemplate.d [hunk 1]
```diff
@@ -1928,6 +1928,10 @@ extern (C++) final class TemplateDeclaration : ScopeDsymbol
                             {
                                 // Allow conversion from T[lwr .. upr] to ref T[upr-lwr]
                             }
+                            else if (global.params.rvalueRefParam)
+                            {
+                                // Allow implicit conversion to ref
```
Comment (12345swordy) [inline src/dmd/dtemplate.d:1933]: Wait hold on, What type of implicit conversion are we talking about here? Non-const Ref parameters shouldn't accept rvalues from implicit conversions here!

**Review by 12345swordy — CHANGES_REQUESTED** [review]
Don't let non-const ref parameters accept rvalues from implicit conversions. Be sure to create a failing test on this.

File: src/dmd/dtemplate.d (hunk 1 above)
Comment (thewilsonator) [reply src/dmd/dtemplate.d:1933]:
```
ref int foo();
void goo(int);
void test()
{
    goo(foo());
}
```
with `-preview=rvaluerefparam` they do and should.

Comment (12345swordy) [reply src/dmd/dtemplate.d:1933]: I am talking about a case scenario involving this:
```int x = 0;
void goo(ref long x) { x = 1} ;
void test()
{
    goo(x); /// a rvalue that is the result of a implicit conversion from an lvalue, which is a bad idea.
    assert(x == 1); // This assertion will fail.
}```
This has brought up regarding the rejection of the rvalue dip here.
https://forum.dlang.org/post/q2pcjn$15nq$1@digitalmars.com

Comment (thewilsonator) [reply src/dmd/dtemplate.d:1933]: Note that this only affects the template deduction portion of the overall validity checking, there is still the function argument matching which will still reject this.

File: test/compilable/issue20705.d [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+{
+    int a;
+}
+
+void bar (T) (const ref T arg) {}
+
+void main ()
+{
+    bar(Foo(42));
+}
```
Comment (Geod24) [inline test/compilable/issue20705.d:12]:
```suggestion
// REQUIRED_ARGS: -preview=rvaluerefparam
struct Foo
{
    int[] a;
}

void bar (T) (const ref T arg) {}
T foo (T) (ref T arg) { return arg; }

void main ()
{
    bar(Foo([42]));
    auto x = foo(Bar([42]));
    static assert(is(typeof(x) == const(Foo)));
}
```

I think that's what @12345swordy has in mind. What's the type of a rvalue `ref`, `T` itself or `const T` ?
I think having it `const T` will be surprising to users, but it also matches the feedback in the DIP.

Comment (thewilsonator) [reply test/compilable/issue20705.d:12]: this is a bit mangled, there is no `Bar`, and `bar` returns `void`. Can you redo  the commit suggestion please?

Comment (Geod24) [reply test/compilable/issue20705.d:12]: My bad. It should've been `Foo`.

File: src/dmd/dtemplate.d (hunk 1 above)
Comment (RazvanN7) [reply src/dmd/dtemplate.d:1933]: Maybe this test should be added to make sure something doesn't brake.

```d
void goo()(ref long x) { x = 1; }
void test()
{
    int x;
    // this should fail.
    goo(x); // a rvalue that is the result of a implicit conversion from an lvalue, which is a bad idea.
    assert(x == 1);
}
```

**Review by 12345swordy — APPROVED** [review]

**Comment by 12345swordy** [comment]
Test are failing for some reason.

**Pull request merged**